cypher/cypher-docs/src/docs/graphgists/uniqueness/uniqueness.asciidoc

```diff
@@ -1,34 +1,107 @@
 
-As can be seen in the above examples, in the case of variable length pattern expressions, it is particularly important to have a uniqueness check, or an infinite number of result records could be found.
+While pattern matching, **Neo4j Cypher** makes sure to not include matches where the same graph relationship is found multiple times in a single path matching a particular pattern.
 
-Cypher makes use of _relationship isomorphism_.
-For example: looking for a user's friends of friends should not return said user.
+Neo4j Cypher makes use of **relationship isomorphism** for path matching and is a very effective way of reducing the result set size and preventing infinite traversals.
+
+[NOTE]
+====
+In Neo4j all relationships have a direction.
+However, you can have the notion of undirected relationships at query time.
+====
+
+In the case of variable length pattern expressions, it is particularly important to have a constraint check, or an infinite number of result records could be found.
+
+To understand this better, let us consider a few alternative options:
+
+
+== Homomorphism
+
+**Constraints:** No constraints for path matching.
+
+
+.Homomorphism
+====
+The graph is composed of only two nodes `(a)` and `(b)`, connected by one relationship, `(a)&#x2D;&#x2D;&#x3E;(b)`.
+
+If the query is looking for paths of length `n` and do not care about the direction, a path of length `n` will be returned repeating the two nodes over and over.
+
+For example, find all paths with 5 relationships and do not care about the relationship direction.
+
+[source, cypher]
+----
+MATCH p=()-[*5]-() RETURN nodes(p)
+----
+
+could return the two resulting records `[a,b,a,b,a,b]` as well as `[b,a,b,a,b,a]`.
+====
+
+
+== Node isomorphism
+
+**Constraints:** The same node cannot be returned more than once for each path matching record.
+
+In the above two-node example, `(a)&#x2D;&#x2D;&#x3E;(b)`, only paths of length 1 can be found with the node isomorphism constraint.
+
+
+.Node isomorphism
+====
+The graph is composed of only two nodes `(a)` and `(b)`, connected by one relationship, `(a)&#x2D;&#x2D;&#x3E;(b)`.
+
+[source, cypher]
+----
+MATCH p=()-[*1]-() RETURN nodes(p)
+----
+
+could return the two resulting records `[a, b]` as well as `[b, a]`.
+====
+
+
+== Relationship isomorphism
+
+**Constraints:** The same relationship cannot be returned more than once for each path matching record.
+
+In the above two-node example, `(a)&#x2D;&#x2D;&#x3E;(b)`, only paths of length 1 can be found with the relationship isomorphism constraint.
+
+
+.Relationship isomorphism
+====
+The graph is composed of only two nodes `(a)` and `(b)`, connected by one relationship, `(a)&#x2D;&#x2D;&#x3E;(b)`.
+
+[source, cypher]
+----
+MATCH p=()-[*1]-() RETURN nodes(p)
+----
+
+could return the two resulting records `[a, b]` as well as `[b, a]`.
+====
+
+
+== Cypher path matching examples
+
+Cypher makes use of relationship isomorphism for path matching.
+
+
+.Friend of friends
```

Comment:
I think this section is a little confusing, with the single example, and numbered sub-queries. I'm not sure of the solution though - I first thought the queries would be better as numbered examples.

However, I wonder whether the layout is better as-is, and just to reconsider the wording around the last query, which is a little confusing comparing back to Query 2.